teos/src/watcher.rs

```diff
@@ -443,127 +372,47 @@ impl Watcher {
         breaches
     }
 
-    /// Filters a map of breaches between those that are valid and those that are not.
+    /// Responds to breaches.
     ///
-    /// Valid breaches are those resulting in a properly formatted [Transaction] once decrypted.
-    fn filter_breaches(
-        &self,
-        breaches: HashMap<Locator, Transaction>,
-    ) -> (
-        HashMap<UUID, Breach>,
-        HashMap<UUID, cryptography::DecryptingError>,
-    ) {
-        let mut valid_breaches = HashMap::new();
-        let mut invalid_breaches = HashMap::new();
+    /// Decrypts triggered appointments using the dispute transaction ID and publishes them.
+    /// If the decryption fails for some appointments or if it succeeds but they get rejected when sent to the network,
+    /// they are marked as an invalid breaches and returned.
+    fn handle_breaches(&self, breaches: HashMap<Locator, Transaction>) -> Vec<UUID> {
+        let mut invalid_breaches = Vec::new();
 
-        // A cache of the already decrypted blobs so replicate decryption can be avoided
-        let mut decrypted_blobs: HashMap<Vec<u8>, Transaction> = HashMap::new();
-
-        let locator_uuid_map = self.locator_uuid_map.lock().unwrap();
-        let dbm = self.dbm.lock().unwrap();
         for (locator, dispute_tx) in breaches.into_iter() {
-            for uuid in locator_uuid_map.get(&locator).unwrap() {
-                let appointment = dbm.load_appointment(*uuid).unwrap();
-                match decrypted_blobs.get(appointment.encrypted_blob()) {
-                    Some(penalty_tx) => {
-                        valid_breaches
-                            .insert(*uuid, Breach::new(dispute_tx.clone(), penalty_tx.clone()));
-                    }
-                    None => {
-                        match cryptography::decrypt(
-                            appointment.encrypted_blob(),
-                            &dispute_tx.txid(),
+            // WARNING(deadlock): Don't lock `self.dbm` over the loop since `Responder::handle_breach` uses it as well.
+            let uuids = self.dbm.lock().unwrap().load_uuids(locator);
+            for uuid in uuids {
+                let appointment = self.dbm.lock().unwrap().load_appointment(uuid).unwrap();
```

Comment:
> What about loading all possible appointments outside the for loop so we don't have to keep holding and dropping the lock over and over?

Acquiring and releasing the lock should be fairly cheap if this is the only load right now. Should be equivalent to ++ then -- operation on an atomic with no waiting at all.
In a normal case with few `uuid`s this shouldn't be bad. But if some attacker is trying to spam the tower with many triggered breaches it would be slow but at least not memory-problematic.